### PR TITLE
improve composer support (no ifndef + autoload.files)

### PR DESCRIPTION
**Pull request by quasilyte**
improve composer support (no ifndef + autoload.files)

* Added comments to the composer-related code

* require on the composer-generated autoload.php does not
  need the #ifndef KPHP anymore

* Added support for autoload.files
  See https://getcomposer.org/doc/04-schema.md#files

When we recognize a require target as a composer-generated
autoload.php file, we handle it in a special way:
all `autoload.files` are required in the place of the initial require.

This transformation can be described like this:

        require_once 'vendor/autoload.php';
        =>
        {
                require_once 'vendor/vkcom/lib/bootstrap.php';
                require_once 'vendor/foo/strings/funcs.php';
                require_once 'vendor/foo/strings/classes.php';
        }

### CODE REVIEW

**Comment by quasilyte** [comment]
It's possible to remove composer-generated files from tests and run `composer install` during the test.

**Comment by quasilyte** [comment]
Why do we need this: some libraries provide functions without classes. They usually provide it via `autoload.files` composer feature. If we support this feature, more 3rd party libraries can be used with KPHP.

Handling `vendor/autoload.php` in a special way also makes it easier to get rid of the `#ifndef KPHP` for composer code as it's very unlikely that we can support genuine composer code in the near future.

File: compiler/composer.h [hunk 1]
```diff
@@ -13,22 +13,48 @@
...
-  void load_root_file(const std::string& filename);
+  // load_file parses the "$pkg_root/composer.json" file and saves
+  // all relevant definitions inside loader;
+  // this method is not thread-safe and should be called only during the compiler init
+  void load_file(const std::string& pkg_root);
+
+  // load_root_file is like load_file, but should be only used
+  // for the main composer file;
+  // this method is not thread-safe and should be called only during the compiler init
+  void load_root_file(const std::string& pkg_root);
```
Comment (AlexK0) [inline compiler/composer.h:33]: Please, check your code style auto-format.

File: compiler/pipes/collect-required-and-classes.cpp [hunk 2]
```diff
@@ -126,6 +127,46 @@ class CollectRequiredPass final : public FunctionPassBase {
...
+      call->location = root->location;
+    }
+    return call;
+  }
+
+  // whether required file name refers to the composer-generated autoload file
+  static bool is_composer_autoload(const std::string &name) {
+    // fast path: it there is no "/autoload.php", then we don't
+    // need to bother and expand the filename
+    if (!vk::contains(name, "/autoload.php")) {
```
Comment (AlexK0) [inline compiler/pipes/collect-required-and-classes.cpp:148]: may be string_view::ends_with is better?

File: compiler/compiler-core.h [hunk 3]
```diff
@@ -49,6 +49,10 @@ class CompilerCore {
   const CompilerSettings &settings() const;
   const string &get_global_namespace() const;
   string unify_file_name(const string &file_name);
+
+  // search_required_file resolves the file_name like it would be expanded when user as require() argument;
+  // it uses search_file_in_include_dirs as well as the current file relative search
+  std::string search_required_file(const std::string &file_name) const;
```
Comment (unserialize) [inline compiler/compiler-core.h:55]: Maybe, move this comment above .cpp implementation?

File: compiler/pipes/collect-required-and-classes.cpp [hunk 4]
```diff
@@ -126,6 +127,46 @@ class CollectRequiredPass final : public FunctionPassBase {
...
+    if (once) {
+      VertexPtr cond = VertexAdaptor<op_log_not>::create(file->get_main_func_run_var());
+      call = VertexAdaptor<op_if>::create(cond, VertexAdaptor<op_seq>::create(call));
+      call->location = root->location;
+    }
+    return call;
+  }
+
+  // whether required file name refers to the composer-generated autoload file
+  static bool is_composer_autoload(const std::string &name) {
```
Comment (unserialize) [inline compiler/pipes/collect-required-and-classes.cpp:144]: Please, rename the parameter name to a more specific: file_name? class_name?

File: compiler/composer.h [hunk 5]
```diff
@@ -13,22 +13,48 @@
 #include "common/mixin/not_copyable.h"
 #include "common/wrappers/string_view.h"
 
+// TODO: reminder: use string_view here instead of const string refs
+
+// TODO: rename to ComposerAutoload?
```
Comment (unserialize) [inline compiler/composer.h:18]: 'ComposerAutoload' seems relevant

File: compiler/pipes/collect-required-and-classes.cpp (hunk 4 above)
Comment (quasilyte) [reply compiler/pipes/collect-required-and-classes.cpp]: Done.

**Comment by quasilyte** [comment]
Addressed review comments.
Rebased on master.

**Comment by quasilyte** [comment]
Failure is related to the PHP and KPHP autoload files evaluation order mismatch.
composer itself gives no control over that order: [one](https://stackoverflow.com/questions/27637114/composer-specify-autoload-files-require-order), [two](https://github.com/composer/composer/issues/6768).
I'll re-check this again and will probably change the tests in a way so they don't rely on the ordering.

**Comment by quasilyte** [comment]
Ready for review.

**Review by AlexK0 — DISMISSED** [review]

**Comment by quasilyte** [comment]
Tests: OK.

**Review by AlexK0 — APPROVED** [review]

**Review by AlexK0 — APPROVED** [review]

**Pull request merged**